sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -446,7 +450,12 @@ case class DataSource(
         dataSource.createRelation(
           sparkSession.sqlContext, mode, caseInsensitiveOptions, Dataset.ofRows(sparkSession, data))
       case format: FileFormat =>
-        sparkSession.sessionState.executePlan(planForWritingFileFormat(format, mode, data)).toRdd
+        val qe = sparkSession.sessionState.executePlan(planForWritingFileFormat(format, mode, data))
+        val insertCommand = qe.executedPlan.collect {
+          case w: WrittenFileCommandExec => w
+        }.head
+        insertCommand.cmd.run(sparkSession, insertCommand.children,
+          metricsCallback.getOrElse(_ => ()))
```

Comment:
This looks like a hack and you did it in 3 different places.
There has to be a way to get it to work via `.execute()` and revert this..